src/librustc_codegen_llvm/debuginfo/metadata.rs

```diff
@@ -1638,10 +1638,10 @@ impl VariantMemberDescriptionFactory<'ll, 'tcx> {
 }
 
 #[derive(Copy, Clone)]
-enum EnumDiscriminantInfo<'ll> {
-    RegularDiscriminant { discr_field: Field, discr_type_metadata: &'ll DIType },
-    OptimizedDiscriminant,
-    NoDiscriminant,
+enum EnumTagInfo<'ll> {
+    RegularTag { tag_field: Field, tag_type_metadata: &'ll DIType },
+    OptimizedTag,
```

Comment:
Does this correspond to `Direct` and `Niche` encodings? If yes, would it make sense to align terminology here?

Reply:
I suspected it might mean something like univariant but I think you're right. Also, instead of `NoTag`, maybe we could use `Option` around `EnumTagInfo`.

Reply:
If we are not sure, maybe I'll just leave this as-is for now and hope someone else cleans it up eventually.^^

Reply:
`OptimizedTag` only refers to `Niche` encodings. `Direct` encodings can be `RegularTag` or `NoTag` depending on whether enum fallback is happening or not. Univariant enums or structs are always `NoTag`.

I do like the idea of using `Option<EnumTagInfo>` and eliminating the `NoTag` variant.

Reply:
r=oli-obk,eddyb with this addressed or a fixme left with instructions on what should be done

Reply:
> Direct encodings can be RegularTag or NoTag depending on whether enum fallback is happening or not. Univariant enums or structs are always NoTag.

What is "enum fallback"? If it's "no tag needed because there is just one variant", AFAIK those do not have a `TagEncoding` at all, they are `Variant::Single`.

Reply:
@oli-obk  I added a FIXME, could you check if that comment makes any sense?

Reply:
> What is "enum fallback"?

It's a windows thing, because windows debuginfo isn't caught up yet